benches/benches.rs

```diff
@@ -0,0 +1,573 @@
+        "../src/interfaces/hybrid_crypto/tests/fin_confidential_user_key.json"
+    ))
+    .expect("cannot deserialize user key JSON");
+    let key_value = &user_decryption_key_json["value"][0]["value"][1]["value"];
+    let hex_key = &key_value[0]["value"]
+        .as_str()
+        .expect("no key as hex found in JSON");
+    let user_decryption_key =
+        UserDecryptionKey::from_bytes(&hex::decode(hex_key).expect("cannot hex decode key"))
+            .expect("cannot generate user private key");
+
+    let mut symmetric_key = vec![0u8; 32];
+    let symmetric_key_ptr = symmetric_key.as_mut_ptr().cast::<i8>();
+    let mut symmetric_key_len = symmetric_key.len() as c_int;
+
+    let mut uid = vec![0u8; 4096];
+    let uid_ptr = uid.as_mut_ptr().cast::<i8>();
+    let mut uid_len = uid.len() as c_int;
+
+    let mut additional_data = vec![0u8; 4096];
+    let additional_data_ptr = additional_data.as_mut_ptr().cast::<i8>();
+    let mut additional_data_len = additional_data.len() as c_int;
+
+    let user_decryption_key_bytes = user_decryption_key
+        .as_bytes()
+        .expect("cannot get bytes from user decryption key");
+    let user_decryption_key_ptr = user_decryption_key_bytes.as_ptr().cast::<i8>();
+    let user_decryption_key_len = user_decryption_key_bytes.len() as i32;
+
+    let mut cache_handle: i32 = 0;
+    unsafe {
+        unwrap_ffi_error(h_aes_create_decryption_cache(
+            &mut cache_handle,
+            user_decryption_key_ptr,
+            user_decryption_key_len,
+        ))
+        .expect("cannot create aes encryption cache");
+    }
+
+    c.bench_function("FFI AES header decryption using cache", |b| {
+        b.iter(|| unsafe {
+            unwrap_ffi_error(h_aes_decrypt_header_using_cache(
+                symmetric_key_ptr,
+                &mut symmetric_key_len,
+                uid_ptr,
+                &mut uid_len,
+                additional_data_ptr,
+                &mut additional_data_len,
+                encrypted_header
+                    .encrypted_header_bytes
+                    .as_ptr()
+                    .cast::<i8>(),
+                encrypted_header.encrypted_header_bytes.len() as c_int,
+                cache_handle,
+            ))
+            .expect("Failed unwrapping FFI AES encrypt header operation")
+        })
+    });
+
+    unsafe {
+        unwrap_ffi_error(h_aes_destroy_decryption_cache(cache_handle))
+            .expect("cannot destroy encryption cache");
+    }
+}
+
+#[cfg(feature = "ffi")]
+unsafe fn unwrap_ffi_error(val: i32) -> anyhow::Result<()> {
+    if val != 0 {
+        let mut message_bytes_key = vec![0u8; 4096];
+        let message_bytes_ptr = message_bytes_key.as_mut_ptr().cast::<i8>();
+        let mut message_bytes_len = message_bytes_key.len() as c_int;
+        get_last_error(message_bytes_ptr, &mut message_bytes_len);
+        let cstr = CStr::from_ptr(message_bytes_ptr);
+        anyhow::bail!("ERROR: {}", cstr.to_str()?);
```

Comment:
You can return an error with `Err(eyre::eyre(...))`. I am not sure this is better, but this would make the two if branch more similar.

Reply:
`bail!` has been designed to avoid writing `Err(eyre::eyre(...))`, plus this will go away when https://github.com/Cosmian/abe_gpsw/issues/9 is done